### PR TITLE
Add new error codes on privacy checking

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @Manishearth

### CODE REVIEW

File: src/librustc_privacy/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(non_snake_case)]
+
+register_long_diagnostics! {
+
+E0445: r##"
+A private trait was used on a "public" type. Erroneous code example:
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: used in the type parameters of a public type

File: src/librustc_privacy/diagnostics.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+```
+trait Foo {
+    fn dummy(&self) { }
+}
+
+pub trait Bar : Foo {} // error: private trait in exported type parameter bound
+```
+
+To solve this error, please ensure the trait is accessible at the same level of
+the type(s) on which it's implemented. Example:
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: types from which it gets exported

File: src/librustc_privacy/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+E0445: r##"
+A private trait was used on a "public" type. Erroneous code example:
+
+```
+trait Foo {
+    fn dummy(&self) { }
+}
+
+pub trait Bar : Foo {} // error: private trait in exported type parameter bound
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: more examples:

`struct Bar<T: Foo>(pub T)`

File: src/librustc_privacy/diagnostics.rs (hunk 1 above)
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: public type parameter bound

File: src/librustc_privacy/diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+mod Foo {
+    struct Bar(u32);
+
+    pub fn bar() -> Bar { // error: private type in exported type signature
+        Bar(0)
+    }
+}
+```
+
+To solve this error, please ensure the type is accessible at the same level of
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: ensure that the type is also public

File: src/librustc_privacy/diagnostics.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/librustc_privacy/diagnostics.rs]: Both of your examples compile... So ?

**Comment by GuillaumeGomez** [comment]
Updated !

**Comment by GuillaumeGomez** [comment]
It should be good (for now !).

File: src/librustc_privacy/diagnostics.rs (hunk 3 above)
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: I didn't put `pub` in my examples

**Comment by GuillaumeGomez** [comment]
I added the other example of erroneous code.

File: src/librustc_privacy/diagnostics.rs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+E0447: r##"
+The `pub` keyword was used inside a function. Erroneous code example:
+
+```
+fn foo() {
+    pub struct Bar; // error: visibility has no effect inside functions
+}
+```
+
+Since we cannot access inside function's elements, the visibility of its
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: we cannot access items defined inside a function, the visibility of its items does ...

File: src/librustc_privacy/diagnostics.rs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    fn foo();
+}
+
+pub impl Bar {} // error: unnecessary visibility qualifier
+
+pub impl Foo for Bar { // error: unnecessary visibility qualifier
+    pub fn foo() {} // error: unnecessary visibility qualifier
+}
+```
+
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: Mention that directly implemented methods share the visibility of the type itself, and trait methods share the visibility of the trait, so `pub` is unnecessary in either case.

File: src/librustc_privacy/diagnostics.rs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+```
+mod Bar {
+    pub struct Foo(isize);
+}
+
+let f = Bar::Foo(0); // error: cannot invoke tuple struct constructor with
+                     //        private fields
+```
+
+To solve this issue, please ensure that all tuple's fields are public. Example:
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: please ensure that all of the fields of the tuple struct are public. Alternatively, provide a `new()` method to the tuple struct to construct it from a given inner value.

```
mod Bar {
  pub struct Foo(isize);
  impl Foo {
   fn new(x: isize) {
     Foo(x)
   }
  }
}
```

File: src/librustc_privacy/diagnostics.rs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        pub a: isize,
+        b: isize,
+    }
+}
+
+let f = Bar::Foo{ a: 0, b: 0 }; // error: field `b` of struct `Bar::Foo`
+                                //        is private
+```
+
+To fix this error, please ensure that all struct's fields are public or please
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: that all the fields of the struct, or implement a function for easy instantiation

File: src/librustc_privacy/diagnostics.rs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+            Foo { a: 0, b: 0 }
+        }
+    }
+}
+
+let f = Bar::Foo::new(); // ok!
+```
+"##,
+
+}
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: newline at end of file

**Comment by GuillaumeGomez** [comment]
Updated !

File: src/librustc_privacy/diagnostics.rs [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+pub impl Foo for Bar { // error: unnecessary visibility qualifier
+    pub fn foo() {} // error: unnecessary visibility qualifier
+}
+```
+
+To fix this error, please remove the visibility qualifier when it is not
+required. Also, directly implemented methods share the visibility of the type
+itself and trait methods share the visibility of the trait, so `pub` is
+unnecessary in either case.
```
Comment (Manishearth) [inline src/librustc_privacy/diagnostics.rs]: This should go before the fix. Explain the problem first, then the solution.

Something like 
"Directly implemented methods .... in either case. 

To fix this error, ..."

**Comment by GuillaumeGomez** [comment]
Updated !

**Pull request merged**